File: Conjugate_gradient/cg.java
public class ConjugateGradientSolver {

    public static void main(String[] args) {
        double[] b = {1.0, 2.0, 3.0};
        double[] x = solve(b, 1e-12, 1000);

        System.out.println("CG Solution:");
        printVector(x);

        // Compute residual A*x - b
        double[] Ax = matrixVectorMultiply(x);
        double[] residual = vectorSubtract(Ax, b);
        System.out.printf("Residual norm: %.2e\n", norm(residual));

        // Exact solution (precomputed)
        double[] xExact = {13.0/28, 6.0/7, 27.0/28};
        System.out.println("Exact solution:");
        printVector(xExact);

        // Compute error
        double[] error = vectorSubtract(x, xExact);
        System.out.printf("Error norm: %.2e\n", norm(error));
    }

    public static double[] solve(double[] b, double tol, int maxIter) {
        int n = b.length;
        double[] x = new double[n];
        double[] r = b.clone();
        double[] p = r.clone();
        double rsold = dotProduct(r, r);

        for (int iter = 0; iter < maxIter; iter++) {
            double[] Ap = matrixVectorMultiply(p);
            double alpha = rsold / dotProduct(p, Ap);

            // Update solution and residual
            x = vectorAdd(x, vectorScale(p, alpha));
            r = vectorSubtract(r, vectorScale(Ap, alpha));

            double rsnew = dotProduct(r, r);
            if (Math.sqrt(rsnew) < tol) {
                System.out.printf("Converged in %d iterations\n", iter+1);
                break;
            }

            double beta = rsnew / rsold;
            p = vectorAdd(r, vectorScale(p, beta));
            rsold = rsnew;
        }
        return x;
    }

    // Tridiagonal matrix-vector multiplication (specific to our 3x3 matrix)
    private static double[] matrixVectorMultiply(double[] p) {
        return new double[] {
            4*p[0] - p[1],
            -p[0] + 4*p[1] - p[2],
            -p[1] + 4*p[2]
        };
    }

    // Helper functions for linear algebra operations
    private static double dotProduct(double[] a, double[] b) {
        double sum = 0.0;
        for (int i = 0; i < a.length; i++) {
            sum += a[i] * b[i];
        }
        return sum;
    }

    private static double[] vectorAdd(double[] a, double[] b) {
        double[] result = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            result[i] = a[i] + b[i];
        }
        return result;
    }

    private static double[] vectorSubtract(double[] a, double[] b) {
        double[] result = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            result[i] = a[i] - b[i];
        }
        return result;
    }

    private static double[] vectorScale(double[] v, double scalar) {
        double[] result = new double[v.length];
        for (int i = 0; i < v.length; i++) {
            result[i] = v[i] * scalar;
        }
        return result;
    }

    private static double norm(double[] v) {
        return Math.sqrt(dotProduct(v, v));
    }

    private static void printVector(double[] v) {
        for (double d : v) {
            System.out.printf("%.12f ", d);
        }
        System.out.println();
    }
}

File: Conjugate_gradient/cg.f
program conjugate_gradient
  implicit none
  integer, parameter :: n = 3
  real(8) :: b(n) = [1.0d0, 2.0d0, 3.0d0]
  real(8) :: x(n), r(n), p(n), Ap(n)
  real(8) :: rsold, rsnew, alpha, beta, tol
  integer :: iter, max_iter = 1000

  ! Initialize solution vector
  x = 0.0d0

  ! Initial residual (r = b - A*x, x=0)
  r = b
  p = r
  rsold = dot_product(r, r)

  tol = 1.0d-6

  ! Conjugate Gradient iterations
  do iter = 1, max_iter
    call matrix_vector_mult(p, Ap)
    alpha = rsold / dot_product(p, Ap)
    
    ! Update solution and residual
    x = x + alpha * p
    r = r - alpha * Ap
    
    rsnew = dot_product(r, r)
    if (sqrt(rsnew) < tol) then
      print *, "Converged in ", iter, " iterations"
      exit
    end if
    
    beta = rsnew / rsold
    p = r + beta * p
    rsold = rsnew
  end do

  if (iter > max_iter) print *, "Did not converge"

  ! Output results
  print *, "CG Solution: ", x

  ! Validate solution
  call matrix_vector_mult(x, Ap)
  r = Ap - b
  print *, "Residual norm: ", sqrt(dot_product(r, r))

  ! Exact solution (precomputed)
  real(8) :: x_exact(n) = [13.0d0/28.0d0, 6.0d0/7.0d0, 27.0d0/28.0d0]
  print *, "Exact solution: ", x_exact
  print *, "Error norm: ", sqrt(dot_product(x - x_exact, x - x_exact))

contains
  ! Tridiagonal matrix-vector multiplication (optimized for this specific matrix)
  subroutine matrix_vector_mult(vec_in, vec_out)
    real(8), intent(in)  :: vec_in(n)
    real(8), intent(out) :: vec_out(n)
    
    vec_out(1) = 4.0d0 * vec_in(1) - vec_in(2)
    vec_out(2) = -vec_in(1) + 4.0d0 * vec_in(2) - vec_in(3)
    vec_out(3) = -vec_in(2) + 4.0d0 * vec_in(3)
  end subroutine matrix_vector_mult
end program conjugate_gradient
